File: exercism/clojure/robot-simulator/src/robot_simulator.clj
(ns robot-simulator)

(defn robot
  [position direction]
  {:coordinates position :bearing direction})

;;https://clojuredocs.org/clojure.core/case
(defn turn-right [direction]
  (case direction
    :east :south
    :south :west
    :west :north
    :north :east))

(defn turn-left [direction]
  (case direction
    :east :north
    :south :east
    :west :south
    :north :west))

(defn- apply-action
  [action robbie]
  (let [
        robbie-coordinates (:coordinates robbie)
        robbie-direction (:bearing robbie)
        ]
    (cond
      (= \A action) (case (:bearing robbie)
                      :north (assoc robbie  :coordinates (assoc robbie-coordinates :y (inc (:y robbie-coordinates))))
                      :south (assoc robbie  :coordinates (assoc robbie-coordinates :y (dec (:y robbie-coordinates))))
                      :east (assoc robbie  :coordinates (assoc robbie-coordinates :x (inc (:x robbie-coordinates))))
                      :west (assoc robbie  :coordinates (assoc robbie-coordinates :x (dec (:x robbie-coordinates)))))
      
      (= \L action) (assoc robbie :bearing (turn-left robbie-direction))
      (= \R action) (assoc robbie :bearing (turn-right robbie-direction))
      )))

(defn simulate [actions robbie]
(cond 
  (empty? actions) robbie
  :else (simulate (rest actions) (apply-action (first actions) robbie))
  ))

File: exercism/julia/robot-simulator/robot-simulator.jl
@enum DIRECTION NORTH=1 WEST=2 EAST=3 SOUTH=4


struct Point
    x::Int
    y::Int
end

mutable struct Robot
    x::Int
    y::Int
    direction::DIRECTION
    Robot(p, d) = new(p[1], p[2], d)
end

function position(r::Robot)
   Point(r.x, r.y)
end 

function heading(r::Robot)
    r.direction
end

const nextdirectionright = Dict(NORTH => EAST, WEST => NORTH, EAST => SOUTH, SOUTH => WEST)
function turn_right!(r::Robot)
    r.direction = nextdirectionright[r.direction]
    r
end
const nextdirectionleft = Dict(NORTH => WEST, WEST => SOUTH, EAST => NORTH, SOUTH => EAST)
function turn_left!(r::Robot)
    r.direction = nextdirectionleft[r.direction]
    r
end

function advance!(r::Robot)
    if r.direction == NORTH
        r.y += 1
    elseif r.direction == WEST
        r.x -= 1
    elseif r.direction == SOUTH
        r.y -= 1
    else
        r.x += 1
    end
    return r
end

const actionToFunction = Dict('A' => advance!, 'R' => turn_right!, 'L' => turn_left!)
function move!(r::Robot, actions::AbstractString)
    for action in actions
        actionToFunction[action](r)
    end
    return r
end
    

